core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -373,8 +373,14 @@ private[spark] class ExecutorAllocationManager(
     // If our target has not changed, do not send a message
     // to the cluster manager and reset our exponential growth
     if (delta == 0) {
-      numExecutorsToAdd = 1
-      return 0
+      // Check if there is any speculative jobs pending
+      if (listener.pendingTasks == 0 && listener.pendingSpeculativeTasks > 0) {
+        numExecutorsTarget =
+          math.max(math.min(maxNumExecutorsNeeded + 1, maxNumExecutors), minNumExecutors)
```

Comment:
@janewangfb Would you please explain why here `+ 1` if there's pending speculativeTasks, should the number of executors be calculated based on the number of pending tasks? Thanks!

Reply:
Same here. 
`maxNumExecutorsNeeded + 1` doesn't quite make sense.
@janewangfb could you please post/update some comments here? And I wonder why we didn't take `pendingSpeculativeTasks` into account when calculating `maxNumExecutorsNeeded()`

Or @jerryshao, do you know the rationale?

Reply:
Also confused by `+1` here. And I think we have already task `pendingSpeculativeTasks` into account @advancedxy :

```
def totalPendingTasks(): Int = {
  pendingTasks + pendingSpeculativeTasks
}
```
Seems this check is redundant.
And it doesn't sync to CM if `numExecutorsTarget ` change(after `+1`).

Reply:
@janewangfb  @cloud-fan Sorry, I realize this is very old PR but found it because I was confused by this logic as well, is there a reason we are adding 1 here with speculative?

Reply:
I can't remember the details as it's too old. But when I look at it again now, this looks a mistake to me: the `+1` seems to try to match the `numExecutorsToAdd = 1` in the previous code. However, `numExecutorsToAdd = 1` doesn't mean we want to allocate one more executor right now.

Reply:
thanks, filed https://issues.apache.org/jira/browse/SPARK-28403